### PR TITLE
Centrally handle user script errors

### LINKED ISSUE #8221: [NBug] The system cannot find the file specified
## Current behaviour

GitExtensions crashes.

## Expected behaviour

GitExtensions shows an error message telling the user about a script that cannot be executed.

## Steps to reproduce
Go to Settings -> Scripts and add a new entry that points to a non-existing executable (in my case it was a ConEmu installation that used to be available at _Program Files (x86)_ and it was replaced with a path pointing to somewhere within _Program Files_).
Execute the script via context menu.

## Error Details
```
System.ComponentModel.Win32Exception (0x80004005): The system cannot find the file specified
   at System.Diagnostics.Process.StartWithCreateProcess(ProcessStartInfo startInfo)
   at System.Diagnostics.Process.Start()
   at GitCommands.Executable.ProcessWrapper..ctor(String fileName, String arguments, String workDir, Boolean createWindow, Boolean redirectInput, Boolean redirectOutput, Encoding outputEncoding)
   at GitCommands.Executable.Start(ArgumentString arguments, Boolean createWindow, Boolean redirectInput, Boolean redirectOutput, Encoding outputEncoding)
   at GitUI.Script.ScriptRunner.RunScript(IWin32Window owner, IGitModule module, ScriptInfo scriptInfo, IGitUICommands uiCommands, RevisionGridControl revisionGrid, Action`1 showError)
   at GitUI.Script.ScriptRunner.RunScript(IWin32Window owner, IGitModule module, String scriptKey, IGitUICommands uiCommands, RevisionGridControl revisionGrid, Action`1 showError)
   at GitUI.Script.ScriptRunner.RunScript(IWin32Window owner, IGitModule module, String scriptKey, IGitUICommands uiCommands, RevisionGridControl revisionGrid)
   at GitUI.Script.ScriptRunner.ExecuteScriptCommand(IWin32Window owner, GitModule module, Int32 command, IGitUICommands uiCommands, RevisionGridControl revisionGrid)
   at GitUI.GitModuleForm.ExecuteCommand(Int32 command)
   at GitUI.CommandsDialogs.FormBrowse.ExecuteCommand(Int32 cmd)
   at ResourceManager.GitExtensionsFormBase.ProcessCmdKey(Message& msg, Keys keyData)
   at System.Windows.Forms.Control.ProcessCmdKey(Message& msg, Keys keyData)
   at System.Windows.Forms.ContainerControl.ProcessCmdKey(Message& msg, Keys keyData)
   at System.Windows.Forms.Control.PreProcessMessage(Message& msg)
   at System.Windows.Forms.Control.PreProcessControlMessageInternal(Control target, Message& msg)
   at System.Windows.Forms.Application.ThreadContext.PreTranslateMessage(MSG& msg)
```

## Additional information

None.

## Environment
- Git Extensions 3.4.1.9675
- Build f49b4d059c5b135b4b2bf9eae4f443f19f6ac5df (Dirty)
- Git 2.26.2.windows.1
- Microsoft Windows NT 10.0.19041.0
- .NET Framework 4.8.4180.0
- DPI 96dpi (no scaling)

### PR DESCRIPTION
**Pull request by RussKie**
Addendum to #8182
Resolves #8221

## Proposed changes

- Centrally handle user script errors

### CODE REVIEW

File: GitUI/Script/ScriptRunner.cs [hunk 1]
```diff
@@ -31,50 +31,53 @@ public static CommandStatus ExecuteScriptCommand(IWin32Window owner, GitModule m
...
+            try
+            {
+                return RunScript(owner, module, scriptKey, uiCommands, revisionGrid,
+                    msg => MessageBox.Show(owner, msg, Strings.Error, MessageBoxButtons.OK, MessageBoxIcon.Error));
+            }
+            catch (Exception ex)
+            {
+                MessageBox.Show(owner,
+                    $"Failed to execute '{scriptKey}' script.{Environment.NewLine}Reason: {ex.Message}",
+                    Strings.Error, MessageBoxButtons.OK, MessageBoxIcon.Error);
```
Comment (mstv) [inline GitUI/Script/ScriptRunner.cs:43]: This should re-use a nearly identical new method of `MessageBoxes` from another open PR. (can be done in a follow-up)

File: IntegrationTests/UI.IntegrationTests/Script/ScriptRunnerTests.cs [hunk 2]
```diff
@@ -24,6 +26,8 @@ public class ScriptRunnerTests
         // Created once for the fixture
         private ReferenceRepository _referenceRepository;
 
+        private static MethodInfo s_miRunScript = typeof(ScriptRunner).GetMethod("RunScript", BindingFlags.NonPublic | BindingFlags.Static);
```
Comment (mstv) [inline IntegrationTests/UI.IntegrationTests/Script/ScriptRunnerTests.cs:29]: 😉 Have you coded too much in C++ recently?
```suggestion
        // new kind of test accessor in Hungarian Notation
        private static MethodInfo s_miRunScript = typeof(ScriptRunner).GetMethod("RunScript", BindingFlags.NonPublic | BindingFlags.Static);
```
Anyways, a comment would have helped me and will do other readers.

File: IntegrationTests/UI.IntegrationTests/Script/ScriptRunnerTests.cs [hunk 3]
```diff
@@ -49,6 +53,12 @@ public void Setup()
             _exampleScript = ScriptManager.GetScript(_keyOfExampleScript);
             _exampleScript.AskConfirmation = false; // avoid any dialogs popping up
             _exampleScript.RunInBackground = true; // avoid any dialogs popping up
+
+            if (s_miRunScript == null)
+            {
+                // s_miRunScript = typeof(ScriptRunner).GetMethod("RunScript", BindingFlags.NonPublic | BindingFlags.Static);
```
Comment (mstv) [inline IntegrationTests/UI.IntegrationTests/Script/ScriptRunnerTests.cs:59]:
```suggestion
```

File: GitUI/Script/ScriptRunner.cs (hunk 1 above)
Comment (RussKie) [reply GitUI/Script/ScriptRunner.cs:43]: I wanted to keep it as small as possible due to plans to push it into 3.4.2.
Agree on a centralised message handling point, as well as pushing the strings out to be localisable. I plan to do this in the master for 4.0.

File: IntegrationTests/UI.IntegrationTests/Script/ScriptRunnerTests.cs (hunk 3 above)
Comment (RussKie) [reply IntegrationTests/UI.IntegrationTests/Script/ScriptRunnerTests.cs:59]: Opps, thanks

File: IntegrationTests/UI.IntegrationTests/Script/ScriptRunnerTests.cs (hunk 2 above)
Comment (RussKie) [reply IntegrationTests/UI.IntegrationTests/Script/ScriptRunnerTests.cs:29]: Hahah, nope, I can barely read C++ let alone write anything in it.
But... I've been spending a lot of time dealing with tests (mostly instabilities 😞), and my colleague developed (yet another!)  test accessor pattern (https://github.com/dotnet/winforms/pull/3225) that our codebase will benefit from too.

As far as notation is concerned - it is what we use in dotnet repos, `s_` indicates static fields.

Comment (RussKie) [reply IntegrationTests/UI.IntegrationTests/Script/ScriptRunnerTests.cs:31]: I put a different comment, please let me know if it needs further clarification

Comment (mstv) [reply IntegrationTests/UI.IntegrationTests/Script/ScriptRunnerTests.cs:31]: With the prefix `s_`, you preach to the converted as for me -- also with `s_mi`. Though the C#/.net coding styles frowned upon.
The suggested comment was in good fun. Your comment is good. I understood the code without it. But it took some time to get the intention "being able to declare the tested method `private`".

Comment (RussKie) [reply IntegrationTests/UI.IntegrationTests/Script/ScriptRunnerTests.cs:31]: I plan to port more dotnet goodness when we move further in to .NET Core :)

**Pull request merged**